protocol-units/dispute/test/RStarM.t.sol

```diff
@@ -0,0 +1,94 @@
+import {
+    IRiscZeroVerifier,
+    Output,
+    OutputLib,
+    Receipt as RiscZeroReceipt,
+    ReceiptClaim,
+    ReceiptClaimLib,
+    ExitCode,
+    SystemExitCode
+} from "../src/IRiscZeroVerifier.sol";
+import {TestReceipt} from "./TestReceipt.sol";
+import {ControlID} from "../src/groth16/ControlID.sol";
+import "forge-std/console2.sol";
+
+contract RStartM is DSTest {
+    using OutputLib for Output;
+    using ReceiptClaimLib for ReceiptClaim;
+
+    Vm vm = Vm(HEVM_ADDRESS);
+    RStarM rStarM;
+    address signer1 = address(0x1);
+    address signer2 = address(0x2);
+    bytes exampleProofData = "exampleProof";
+    address validator = address(0x1);
+
+    RiscZeroReceipt internal TEST_RECEIPT = RiscZeroReceipt(
+        TestReceipt.SEAL,
+        ReceiptClaim(
+            TestReceipt.IMAGE_ID,
+            TestReceipt.POST_DIGEST,
+            ExitCode(SystemExitCode.Halted, 0),
+            bytes32(0x0000000000000000000000000000000000000000000000000000000000000000),
+            Output(sha256(TestReceipt.JOURNAL), bytes32(0)).digest()
+        )
+    );
+
+    function setUp() public {
+        rStarM = new RStarM(1, 15, 2, ControlID.CONTROL_ID_0, ControlID.CONTROL_ID_1, ControlID.BN254_CONTROL_ID);
+        vm.deal(address(this), rStarM.MIN_STAKE());
+    }
+
+    function testRegisterValidator() public {
+      uint256 initialBalance = 666 ether;
+      vm.deal(signer1, initialBalance);
+      uint256 minStake = rStarM.MIN_STAKE();
+
+      vm.prank(signer1);
+      rStarM.registerValidator{value: minStake}();
+
+      (bool isRegistered, uint256 stake) = rStarM.validators(signer1);
+      assertTrue(isRegistered, "Validator should be registered");
+      assertEq(stake, minStake, "Validator stake should match the provided stake");
+    } 
+
+    function testVerifyKnownGoodReceipt() external view {
+        require(rStarM.verify_integrity(TEST_RECEIPT), "verification failed");
+    }
+
+    function testVerifyKnownGoodImageIdAndJournal() external view {
+        require(
+            rStarM.verify(
+                TEST_RECEIPT.seal, TestReceipt.IMAGE_ID, TEST_RECEIPT.claim.postStateDigest, sha256(TestReceipt.JOURNAL)
+            ),
+            "verification failed"
+        );
+    }
+
+    function testVerifyMangledReceipts() external view {
+        RiscZeroReceipt memory mangled = TEST_RECEIPT;
+
+        mangled.seal[0] ^= bytes1(uint8(1));
+        require(!rStarM.verify_integrity(mangled), "verification passed on mangled seal value");
+        mangled = TEST_RECEIPT;
+
+        console2.log("verified!");
+        mangled.claim.postStateDigest ^= bytes32(uint256(1));
+        require(!rStarM.verify_integrity(mangled), "verification passed on mangled postStateDigest value");
+        mangled = TEST_RECEIPT;
+
+        mangled.claim.output ^= bytes32(uint256(1));
+        require(!rStarM.verify_integrity(mangled), "verification passed on mangled input value");
+        mangled = TEST_RECEIPT;
+    }
```

Comment:
I'd like to see some tests on the optimistic commitment. Include cases for fraudulent actors, honest actors, a mix, etc.